### DISCUSSION TITLE
Help Request - Extending the Neovim Extendability Documentation

### DISCUSSION THREAD

**Discussion by tacocats** [neovim/neovim#26683] [General]
Hi, I’m trying to compile a general doc with all the ways to extend Neovim, while there’s a lot of documentation out there I feel it’s sort of fragmented. Here’s a general summary of what I’ve collected so far. If anyone knows of anything else or if I misrepresented anything, please let me know and I’ll update this. I’d like to see if I can get a page added to Neovim wiki sort of as a general overview so we can keep track of all the ways to Neovim can be extended.

Here's a rough draft of what I have so far:
**NeoVim Plugin Language Options**

* VimScript
* Lua
* Providers - see `:h provider`
* RemotePlugins - By using RPC any language can be used
* WASM? - While not supported, this would allow for plugins to be sandboxed and allow extending neovim with multiple languages through a single API. I only mention this since I saw a few issues opened about it and it sounded like this was a possibility in the future, what are your thoughts on removing this?
* Lua Transpilers - Lua has a few Transpilers, these are languages that will convert to Lua.
  * Some are listed here - [**https://github.com/neovim/neovim/wiki/Related-projects#transpilers**](https://github.com/neovim/neovim/wiki/Related-projects#transpilers)
* Vim9 script - Not supported, probably never will be from my understanding. This was implemented in Vim9 release. This was a new language implemented in VIm9 to speed up the language and try to make it more modern. It currently only works in Vim9+

**APIS**

* RPC/Remote plugins can use - https://neovim.io/doc/user/api.html#api
  * There’s client setup for multiple languages to aid in this process - https://github.com/neovim/neovim/wiki/Related-projects#api-clients
* Providers API - Only a few languages are supported, from my understand this is the old way of doing it from VIM before RPC see `:help provider` for more information
* Lua API 
* VimScript API (eval) - https://neovim.io/doc/user/eval.html#functions
* LUA STDLIB - Since it’s Lua you have access to the whole language here https://www.lua.org/manual/5.1/

**Comment by justinmk** [comment]
Looks fairly complete, though there is also:

- luajit FFI
- load `libnvim` as a C module (unsupported, but works)

Only Lua and RPC API are recommended, so the "fragmentation" is simply the result of the fact that none of the other things are recommended.

**Reply by HiPhish** [reply]
> Only Lua and RPC API are recommended

Wait, are remote plugins deprecated or something? I think they are really great for when we really need some library from another language. They might not be as popular as in the past now that we have Lua, but they are still useful.

**Reply by clason** [reply]
Remote plugins _are_ "RPC API".

**Reply by HiPhish** [reply]
Ah OK then :) Well, there is a bit more machinery (`:UpdateRemotePlugins`) which I was afraid might be marked for deprecation or something like that.
